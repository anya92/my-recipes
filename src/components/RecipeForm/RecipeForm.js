import React, { Component } from 'react';
import uniqId from 'uniqid';
import getSlug from '../../helpers/slug';

import { Input, Textarea } from './FormFields/InputAndTextarea';
import Checkbox from './FormFields/Checkbox';
import Select from './FormFields/Select';
import Ingredients from './FormFields/Ingredients';
import ImageForm from './FormFields/ImageForm';

import { Form, Row } from './styled/Form';
import Button from './styled/Button';

export class RecipeForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: props.recipe ? props.recipe.name : '',
      directions: props.recipe ? props.recipe.directions : '',
      ingredients: props.recipe ? props.recipe.ingredients : [{name: '', quantity: ''}, {name: '', quantity: ''}, {name: '', quantity: ''}],
      category: props.recipe ? props.recipe.category : '',
      tags: props.recipe ? props.recipe.tags : [],
      time: props.recipe ? props.recipe.time : '',
      yields: props.recipe ? props.recipe.yields : '',
      image: props.recipe ? props.recipe.image : '',
      error: '',
      errors: {
        name: false,
        directions: false,
        ingredients: false,
        category: false,
        time: false,
        yields: false,
      },
    };
  }

  onInputChange = (e, type) => {
    const { value } = e.target;
    this.setState(() => ({ [type]: value }));
  }

  setImage = image => {
    this.setState(() => ({ image }));
  }

  onIngredientChange = (e, type, index) => {
    const { value } = e.target;
    const ingredients = this.state.ingredients.map((ingredient, i) => {
      if (i === index) {
        ingredient[type] = value;
      }
      return ingredient;
    });
    this.setState(() => ({ ingredients }));
  }

  addIngredientInput = () => {
    this.setState(prevState => {
      return {
        ingredients: [
          ...prevState.ingredients,
          { name: '', quantity: '' }
        ]
      };
    });
  }

  removeIngredientInput = index => {
    if (this.state.ingredients.length > 1) {
      this.setState(prevState => {
        return {
          ingredients: [
            ...prevState.ingredients.slice(0, index),
            ...prevState.ingredients.slice(index + 1)
          ]
        };
      });
    }
  }

  onTagChange = tag => {
    if (!this.state.tags.includes(tag)) {
      this.setState(prevState => ({ tags: [...prevState.tags, tag] }));
    } else {
      const index = this.state.tags.indexOf(tag);
      this.setState(prevState => ({ tags: [...prevState.tags.slice(0, index), ...prevState.tags.slice(index + 1)] }));
    }
  }

  validateFields = (field) => {
    if (field === 'ingredients') {
      if (!this.state.ingredients.some(ingredient => ingredient.name !== '')) {
        this.setState(prevState => ({ errors: { ...prevState.errors, ingredients: true } }));
        return false;
      } else {
        this.setState(prevState => ({ errors: { ...prevState.errors, ingredients: false } }));
        return true;
      }
    }
    if (!this.state[field]) {
      this.setState(prevState => ({ errors: { ...prevState.errors, [field]: true } }));
      return false;
    } else {
      this.setState(prevState => ({ errors: { ...prevState.errors, [field]: false } }));
      return true;
    }
  }

  onSubmit = e => {
    e.preventDefault();
    // format (= delete empty ingredients)
    const ingredients = this.state.ingredients.filter(ingredient => (
      ingredient.name !== ''
    ));
    // validate required fields
    const requiredFields = ['name', 'directions', 'ingredients', 'category', 'time', 'yields'];

    if (!requiredFields.every(field => this.validateFields(field))) {
      this.setState(() => ({ error: 'Please provide all requested information (*)' }));
    } else {
      // save a recipe
      this.setState(() => ({ error: '' }));
      const { name, directions, category, tags, time, yields, image } = this.state;
      const { recipe, recipes } = this.props;
      const id = recipe ? recipe.id : uniqId();
      const addedAt = recipe ? recipe.addedAt : Date.now();
      const slug = getSlug(recipe, name, recipes);
      this.props.onSubmit({ id, addedAt, slug, name, directions, ingredients, category, tags, time, yields, image });
    }
  }

  render() {
    return (
      <Form onSubmit={this.onSubmit}>
        <div>
          <Input 
            name="name" 
            value={this.state.name} 
            handleInputChange={this.onInputChange} 
            error={this.state.errors.name}
            isRequired={true}
          />
          <Textarea 
            name="directions" 
            value={this.state.directions} 
            handleInputChange={this.onInputChange} 
            error={this.state.errors.directions}
            isRequired={true}
          />
          <Ingredients 
            ingredients={this.state.ingredients}
            handleIngredientChange={this.onIngredientChange}
            addIngredientInput={this.addIngredientInput}
            removeIngredientInput={this.removeIngredientInput}
            error={this.state.errors.ingredients}
          />
          <Select 
            category={this.state.category} 
            handleInputChange={this.onInputChange} 
            error={this.state.errors.category}
          />
          <Row>
            <Input 
              name="time" 
              type="number"
              value={this.state.time} 
              handleInputChange={this.onInputChange} 
              error={this.state.errors.time} 
              additionalText="minute/s"
              isRequired={true}
            />
            <Input 
              name="yields" 
              value={this.state.yields} 
              handleInputChange={this.onInputChange} 
              error={this.state.errors.yields} 
              additionalText="serving/s"
              isRequired={true}
            />          
          </Row>  
        </div>
        <div>
          <ImageForm 
            image={this.state.image} 
            setImage={this.setImage} 
          />
          <Checkbox 
            tags={this.state.tags} 
            handleTagChange={this.onTagChange} 
          />
          { this.state.error && <p>{this.state.error}</p> } {/* Error */}
          <Row>
            <Button type="button" onClick={this.props.onCancel}>Cancel</Button>
            <Button type="submit" blue>Save</Button>
          </Row>  
        </div>
      </Form>
    );
  }
}

export default RecipeForm;
