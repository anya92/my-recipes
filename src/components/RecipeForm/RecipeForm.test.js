import React from 'react';
import { shallow } from 'enzyme';
import recipes from '../../recipes';

import RecipeForm from './RecipeForm';

let wrapper;

beforeEach(() => {
  wrapper = shallow(<RecipeForm />);
});

test('should render RecipeForm correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render RecipeForm with data', () => {
  wrapper = shallow(<RecipeForm recipe={recipes[0]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should set name on input change', () => {
  const value = 'Spaghetti';
  wrapper.find('Input').at(0).prop('handleInputChange')({ target: { value } }, 'name');
  expect(wrapper.state('name')).toBe(value);
});

test('should set directions on textarea change', () => {
  const value = 'New directions';
  wrapper.find('Textarea').prop('handleInputChange')({ target: { value } }, 'directions');
  expect(wrapper.state('directions')).toBe(value);
});

test('should set ingredients on change', () => {
  const name = 'eggs';
  const quantity = '2-4';
  wrapper.find('Ingredients').prop('handleIngredientChange')({ target: { value: name } }, 'name', 0);
  wrapper.find('Ingredients').prop('handleIngredientChange')({ target: { value: quantity } }, 'quantity', 0);
  expect(wrapper.state('ingredients')[0]).toEqual({ name, quantity });
});

test('should add new ingredient input on addIngredientInput', () => {
  wrapper.find('Ingredients').prop('addIngredientInput')();
  expect(wrapper.state('ingredients').length).toBe(4);
});

test('should remove ingredient input on removeIngredientInput', () => {
  wrapper.find('Ingredients').prop('removeIngredientInput')(2);
  expect(wrapper.state('ingredients').length).toBe(2);
});

test('should set category on select change', () => {
  const value = 'breakfast';
  wrapper.find('Select').prop('handleInputChange')({ target: { value } }, 'category');
  expect(wrapper.state('category')).toBe(value);
});

test('should set tag on checkbox change', () => {
  wrapper.find('Checkbox').prop('handleTagChange')('vegan');
  expect(wrapper.state('tags')).toContain('vegan');
});

test('should remove tag when it is unchecked', () => {
  wrapper = shallow(<RecipeForm recipe={recipes[0]} />);
  wrapper.find('Checkbox').prop('handleTagChange')('healthy');
  expect(wrapper.state('tags')).not.toContain('healthy');
});

test('should set time on change', () => {
  const value = 12;
  wrapper.find('Input').at(1).prop('handleInputChange')({ target: { value } }, 'time');
  expect(wrapper.state('time')).toBe(value);
});

test('should set yields on change', () => {
  const value = '2-4';
  wrapper.find('Input').at(2).prop('handleInputChange')({ target: { value } }, 'yields');
  expect(wrapper.state('yields')).toBe(value);
});

test('should set image', () => {
  const image = 'assets/images/pancakes.jpg';
  wrapper.find('ImageForm').prop('setImage')(image);
  expect(wrapper.state('image')).toBe(image);
});

test('should call onCancel prop on "Cancel" button click', () => {
  const onCancelSpy = jest.fn();
  wrapper = shallow(<RecipeForm onCancel={onCancelSpy} />);
  wrapper.find('Button').at(0).simulate('click');
  expect(onCancelSpy).toHaveBeenCalled();
});

test('should render error for invalid form submission', () => {
  expect(wrapper).toMatchSnapshot();
  wrapper.find('Form').simulate('submit', {
    preventDefault: () => {},
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  wrapper = shallow(<RecipeForm recipe={recipes[0]} onSubmit={onSubmitSpy} />);
  wrapper.find('Form').simulate('submit', {
    preventDefault: () => {},
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith(recipes[0]);
});

