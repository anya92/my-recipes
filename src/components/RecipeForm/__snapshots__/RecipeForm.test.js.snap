// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RecipeForm correctly 1`] = `
<Form
  onSubmit={[Function]}
>
  <div>
    <Input
      additionalText=""
      error={false}
      handleInputChange={[Function]}
      isRequired={true}
      name="name"
      type="text"
      value=""
    />
    <Textarea
      error={false}
      handleInputChange={[Function]}
      isRequired={true}
      name="directions"
      value=""
    />
    <Ingredients
      addIngredientInput={[Function]}
      error={false}
      handleIngredientChange={[Function]}
      ingredients={
        Array [
          Object {
            "name": "",
            "quantity": "",
          },
          Object {
            "name": "",
            "quantity": "",
          },
          Object {
            "name": "",
            "quantity": "",
          },
        ]
      }
      removeIngredientInput={[Function]}
    />
    <Select
      category=""
      error={false}
      handleInputChange={[Function]}
    />
    <styled.div>
      <Input
        additionalText="minute/s"
        error={false}
        handleInputChange={[Function]}
        isRequired={true}
        name="time"
        type="number"
        value=""
      />
      <Input
        additionalText="serving/s"
        error={false}
        handleInputChange={[Function]}
        isRequired={true}
        name="yields"
        type="text"
        value=""
      />
    </styled.div>
  </div>
  <div>
    <ImageForm
      image=""
      setImage={[Function]}
    />
    <Checkbox
      handleTagChange={[Function]}
      tags={Array []}
    />
    <styled.div>
      <Button
        type="button"
      >
        Cancel
      </Button>
      <Button
        blue={true}
        type="submit"
      >
        Save
      </Button>
    </styled.div>
  </div>
</Form>
`;

exports[`should render RecipeForm with data 1`] = `
<Form
  onSubmit={[Function]}
>
  <div>
    <Input
      additionalText=""
      error={false}
      handleInputChange={[Function]}
      isRequired={true}
      name="name"
      type="text"
      value="Fried Eggs"
    />
    <Textarea
      error={false}
      handleInputChange={[Function]}
      isRequired={true}
      name="directions"
      value="Heat butter in nonstick skillet. Break eggs and slip into pan, one at a time. Cook slowly until whites are completely set and yolks begin to thicken but are not hard, 5 to 6 minutes. Slide turner under each egg and carefully flip it over in pan. Cook second side to desired doneness. Sprinkle with salt and pepper. Serve immediately."
    />
    <Ingredients
      addIngredientInput={[Function]}
      error={false}
      handleIngredientChange={[Function]}
      ingredients={
        Array [
          Object {
            "name": "eggs",
            "quantity": "2-4",
          },
          Object {
            "name": "butter",
          },
          Object {
            "name": "salt",
          },
          Object {
            "name": "pepper",
          },
        ]
      }
      removeIngredientInput={[Function]}
    />
    <Select
      category="breakfast"
      error={false}
      handleInputChange={[Function]}
    />
    <styled.div>
      <Input
        additionalText="minute/s"
        error={false}
        handleInputChange={[Function]}
        isRequired={true}
        name="time"
        type="number"
        value="9"
      />
      <Input
        additionalText="serving/s"
        error={false}
        handleInputChange={[Function]}
        isRequired={true}
        name="yields"
        type="text"
        value="2 to 4"
      />
    </styled.div>
  </div>
  <div>
    <ImageForm
      image="assets/images/fried-eggs.jpg"
      setImage={[Function]}
    />
    <Checkbox
      handleTagChange={[Function]}
      tags={
        Array [
          "quick&easy",
          "healthy",
        ]
      }
    />
    <styled.div>
      <Button
        type="button"
      >
        Cancel
      </Button>
      <Button
        blue={true}
        type="submit"
      >
        Save
      </Button>
    </styled.div>
  </div>
</Form>
`;

exports[`should render error for invalid form submission 1`] = `
<Form
  onSubmit={[Function]}
>
  <div>
    <Input
      additionalText=""
      error={false}
      handleInputChange={[Function]}
      isRequired={true}
      name="name"
      type="text"
      value=""
    />
    <Textarea
      error={false}
      handleInputChange={[Function]}
      isRequired={true}
      name="directions"
      value=""
    />
    <Ingredients
      addIngredientInput={[Function]}
      error={false}
      handleIngredientChange={[Function]}
      ingredients={
        Array [
          Object {
            "name": "",
            "quantity": "",
          },
          Object {
            "name": "",
            "quantity": "",
          },
          Object {
            "name": "",
            "quantity": "",
          },
        ]
      }
      removeIngredientInput={[Function]}
    />
    <Select
      category=""
      error={false}
      handleInputChange={[Function]}
    />
    <styled.div>
      <Input
        additionalText="minute/s"
        error={false}
        handleInputChange={[Function]}
        isRequired={true}
        name="time"
        type="number"
        value=""
      />
      <Input
        additionalText="serving/s"
        error={false}
        handleInputChange={[Function]}
        isRequired={true}
        name="yields"
        type="text"
        value=""
      />
    </styled.div>
  </div>
  <div>
    <ImageForm
      image=""
      setImage={[Function]}
    />
    <Checkbox
      handleTagChange={[Function]}
      tags={Array []}
    />
    <styled.div>
      <Button
        type="button"
      >
        Cancel
      </Button>
      <Button
        blue={true}
        type="submit"
      >
        Save
      </Button>
    </styled.div>
  </div>
</Form>
`;

exports[`should render error for invalid form submission 2`] = `
<Form
  onSubmit={[Function]}
>
  <div>
    <Input
      additionalText=""
      error={true}
      handleInputChange={[Function]}
      isRequired={true}
      name="name"
      type="text"
      value=""
    />
    <Textarea
      error={false}
      handleInputChange={[Function]}
      isRequired={true}
      name="directions"
      value=""
    />
    <Ingredients
      addIngredientInput={[Function]}
      error={false}
      handleIngredientChange={[Function]}
      ingredients={
        Array [
          Object {
            "name": "",
            "quantity": "",
          },
          Object {
            "name": "",
            "quantity": "",
          },
          Object {
            "name": "",
            "quantity": "",
          },
        ]
      }
      removeIngredientInput={[Function]}
    />
    <Select
      category=""
      error={false}
      handleInputChange={[Function]}
    />
    <styled.div>
      <Input
        additionalText="minute/s"
        error={false}
        handleInputChange={[Function]}
        isRequired={true}
        name="time"
        type="number"
        value=""
      />
      <Input
        additionalText="serving/s"
        error={false}
        handleInputChange={[Function]}
        isRequired={true}
        name="yields"
        type="text"
        value=""
      />
    </styled.div>
  </div>
  <div>
    <ImageForm
      image=""
      setImage={[Function]}
    />
    <Checkbox
      handleTagChange={[Function]}
      tags={Array []}
    />
    <styled.div>
      Please provide all requested information (*)
    </styled.div>
    <styled.div>
      <Button
        type="button"
      >
        Cancel
      </Button>
      <Button
        blue={true}
        type="submit"
      >
        Save
      </Button>
    </styled.div>
  </div>
</Form>
`;
